src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/haflows/HaFlowHistorySpec.groovy

```diff
@@ -0,0 +1,246 @@
+        }
+
+        cleanup:
+        haFlow && !deletedFlow && haFlow.delete()
+    }
+
+    @Tags(LOW_PRIORITY)
+    @Tidy
+    def "History records are created during link unsuccessful rerouting with root cause details and can be retrieved with or without timeline"() {
+        given: "HA-Flow has been created"
+        def swT = topologyHelper.switchTriplets[0]
+        Long timeBeforeOperation = System.currentTimeSeconds()
+        HaFlowExtended haFlow = HaFlowExtended.build(swT, northboundV2, topology).create()
+
+        when: "Deactivate the shared switch"
+        def blockData = switchHelper.knockoutSwitch(swT.shared, RW)
+
+        and: "Related ISLs are FAILED"
+        def isls = topology.getRelatedIsls(swT.shared)
+        Wrappers.wait(discoveryTimeout + WAIT_OFFSET / 2) {
+            def allIsls = northbound.getAllLinks()
+            isls.each { assert islUtils.getIslInfo(allIsls, it).get().actualState == IslChangeType.FAILED }
+        }
+
+        and: "Ha-Flow goes DOWN"
+        haFlow.waitForBeingInState(FlowState.DOWN)
+
+        then: "Correct event appears in HA-Flow history and can be retrieved without specifying timeline"
+        Wrappers.wait(WAIT_OFFSET) {
+            assert haFlow.getHistory().getEntriesByType(HaFlowActionType.REROUTE_FAIL)[0].payloads.find {
+                it.action == HaFlowActionType.REROUTE_FAIL.payloadLastAction
+            }
+        }
+
+        def historyRecordWithoutTimeline = haFlow.getHistory().getEntriesByType(HaFlowActionType.REROUTE_FAIL)
+        historyRecordWithoutTimeline.size() == 1
+
+        and: "All basic fields are correct and rerouting failure details are available"
+        verifyAll {
+            historyRecordWithoutTimeline[0].haFlowId == haFlow.haFlowId
+            historyRecordWithoutTimeline[0].taskId
+            historyRecordWithoutTimeline[0].timestampIso
+            historyRecordWithoutTimeline[0].payloads.every { it.timestampIso }
+            historyRecordWithoutTimeline.dumps.flatten().isEmpty()
+            historyRecordWithoutTimeline[0].payloads[-1].action == HaFlowActionType.REROUTE_FAIL.getPayloadLastAction()
+
+            historyRecordWithoutTimeline[0].payloads[-1].details == "ValidateHaFlowAction failed: HA-flow's $haFlow.haFlowId src switch ${swT.shared.dpId} is not active"
+        }
+
+        and: "Correct event appears in HA-Flow history and can be retrieved with specifying timeline"
+        def historyRecordsWithTimeline = haFlow.getHistory(timeBeforeOperation, System.currentTimeSeconds())
+                .getEntriesByType(HaFlowActionType.REROUTE_FAIL)
+
+        and: "Both retrieved history records are identical"
+        historyRecordWithoutTimeline == historyRecordsWithTimeline
+
+        cleanup:
+        blockData && switchHelper.reviveSwitch(swT.shared, blockData, true)
+        haFlow && haFlow.delete()
+    }
+
+    @Tidy
+    def "Empty history returned in case filters return no results"() {
```

Comment:
I'd expect this test to be low priority as well.

Reply:
Done